### DISCUSSION TITLE
Tables are not found if their names are in upper case

### DISCUSSION THREAD

**Discussion by informatik01** [dbeaver/dbeaver#22580] [Ideas]
Hi.

When I use "Open Database Object" ( <kbd>⌃</kbd> <kbd>⇧</kbd> <kbd>D</kbd> on Mac) or "Search" (<kbd>⌃</kbd> <kbd>H</kbd> on mac) on my local MySQL database, everything is found just fine. So I can quickly open some table etc.

But when I perform the same actions on my remote databases - nothing gets found except for objects from MySQL's standard  schemas like `sys` and `performance`.

<img width="640" alt="image" src="https://github.com/dbeaver/dbeaver/assets/11339880/fe4f73f8-4788-465e-9f32-c74b84f18967">

---
What is interesting that there is no problem with navigating to the desired tables on the remote servers using the "Database Navigator": all the tables are shown just fine.  Still when trying the above actions (opening objects and searching) - this issue happens.

When using other database clients, like JetBrains DataGrip, TablePlus, DBVisualizer - no problem with opening remote objects. It's only with DBeaver such story occurs.

Any ideas why is it so?

---

### Platform and version info:
- DBeaver Community Edition version 23.3.3.202401211839
- macOS Sonoma version 14.3 (on Apple Silicon)

**Comment by LonwoLonwo** [comment]
Hello @informatik01 

What remote instances do you use? Please provide more info about where they are stored.

**Comment by informatik01** [comment]
@LonwoLonwo 
These are the servers we use at our work. They must not be accessible from "outside" the world, but used internally by our server side services. So I cannot disclose their addresses.

In other words, I connect to MySQL instances which are run NOT locally (i.e. on my machine), but somewhere in our corporate LAN.

It's just really strange that all the schemas and tables are nicely shown in "Database Navigator", but when you try to open or search for them using the above mentioned tools - only the objects (i.e. tables, views etc) from the standard MySQL schemas can be found.

**Reply by LonwoLonwo** [reply]
I can not reproduce your issue, unfortunately.
I have a remote MySQL database with an SSH connection and a GCP MySQL database, and both of them return results in the search dialog.

You can investigate this issue from your side with the [Query Manager](https://github.com/dbeaver/dbeaver/wiki/Query-Manager) (with all enabled filters in their preference page).
If Query Manager shows you non-empty results, you can check your logs and share them with me. 
You can find instructions here: https://dbeaver.com/docs/wiki/Log-files/.

**Comment by informatik01** [comment]
@LonwoLonwo 
Thank you.

Will get back to it later, probably on weekends, at work ATM.

**Comment by informatik01** [comment]
@LonwoLonwo 
Just checked the reported issue following your advice to use "Query manager" and quickly found the reason.

Turned out that our Staging and Production servers are configured with the _system variable_ [`lower_case_table_names`](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_lower_case_table_names) set to zero:
> `lower_case_table_names=0`

(NOTE: on my local Mac machine I have this system variable set to `2`, so no problems with it locally)

If this server system variable is set to `0`, table names are stored _as specified_ and 👉 **comparisons are case-sensitive**. We have all our tables historically defined in UPPER case with underscores (if needed). And I can see in the "Query Manager" logs, that when using "Open Database Object" functionality, it performs **_case insensitive_** search using the following SQL query (example of searching for a table that starts with the letter "k"):
```sql
SELECT TABLE_SCHEMA,
       TABLE_NAME
FROM information_schema.TABLES
WHERE TABLE_NAME LIKE 'k%'
ORDER BY TABLE_NAME
LIMIT 1000
```
Even if I type a search word using upper case letters, the search is still done using the lower case letters. And as a result - nothing (besides MySQL built-in tables) is found.

Can this search pattern be somehow overridden? Or is there any other solution for the above described case?

**Reply by LonwoLonwo** [reply]
According to this [discussion](https://stackoverflow.com/questions/5629111/how-can-i-make-sql-case-sensitive-string-comparison-on-mysql) - does this query work in your case?

```sql
SELECT TABLE_SCHEMA,
       TABLE_NAME
FROM information_schema.TABLES
WHERE TABLE_NAME LIKE BINARY 'k%'
ORDER BY TABLE_NAME
LIMIT 1000
```

**Reply by informatik01** [reply]
@LonwoLonwo 
Nope, it didn't help.

Btw, when executing this SQL on MySQL versions >= 8.0.27, you will get the following warning:
> 'BINARY expr' is **deprecated** and will be removed in a future release. Please use CAST instead

https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html#operator_binary

To fix this issue, we can modify the above SQL expression like this:
```sql
SELECT TABLE_SCHEMA,
       TABLE_NAME
FROM information_schema.TABLES
WHERE TABLE_NAME LIKE CAST('k%' AS BINARY)
ORDER BY TABLE_NAME
LIMIT 1000;
```

Now there will be no warning.

**Reply by LonwoLonwo** [reply]
> Now there will be no warning. 

And no result also?

And what about this version:

```sql
SELECT TABLE_SCHEMA,
       TABLE_NAME
FROM information_schema.TABLES
WHERE UPPER(TABLE_NAME) LIKE 'K%'
ORDER BY TABLE_NAME
LIMIT 1000;
```

**Reply by informatik01** [reply]
@LonwoLonwo 
1) Why no result?
The result will be **the same as previously**, just **no warning anymore**.
What I meant is that it didn't fix my problem. I just wanted to make a notice about how to get get rid of that warning.

2) Please see my last [post below,](https://github.com/orgs/dbeaver/discussions/22580#discussioncomment-8468464) where I proposed two solutions which DO return the desired results.

**Reply by LonwoLonwo** [reply]
Ok, will be fixed here: https://github.com/dbeaver/dbeaver/pull/22901

**Reply by informatik01** [reply]
@LonwoLonwo 
Nice, thank you!

**Comment by informatik01** [comment]
@LonwoLonwo 
Could you please give some feedback on my previous post above?

**Reply by LonwoLonwo** [reply]
Ok, thanks for the excellent explanation.
Do you have any ideas on how to solve it? From your side.

**Comment by informatik01** [comment]
@LonwoLonwo 

From my side - as I already described in my previous post, I cannot _force_ DBeaver to respect case sensitivity.
What I would expect is that DBeaver provides an **option to respect the case of search letters**.

For example, in the built-in Search window ( <kbd>Ctrl</kbd> + <kbd>H</kbd> ) which has 
- File Search
- DB Full-Text
- DB Metadata

\- all these search options have a dedicated checkbox "**Case sensitive**" that allows respect the case of a search letters, e.g.:

![image](https://github.com/dbeaver/dbeaver/assets/11339880/3776b4bc-0a10-47b0-995f-4c4664d043a1)

---
💡 It would be really nice to have the same checkbox in the "Open Database Object" dialog window too, because currently it is missing there:

![image](https://github.com/dbeaver/dbeaver/assets/11339880/63202d84-9eec-451a-9bf2-ff1117359a5b)

---

#### 🙋‍♂️ So could DBeaver users expect such an option in nearest future?

**Comment by informatik01** [comment]
@LonwoLonwo 
Actually, even more simpler (in  the meaning that no need to add extra checkbox) would be to use _more unified_ SQL expression under the hood for the "Open Database Object" functionality.

# Problem
So here is the original SQL expression which is used by DBeaver now for "Open Database Object" functionality:
```sql
SELECT TABLE_SCHEMA,
       TABLE_NAME
FROM information_schema.TABLES
WHERE TABLE_NAME LIKE 'k%'
ORDER BY TABLE_NAME
LIMIT 1000;
```

# Solution
Below are two SQL expressions which both achieve the desired result for the reported case.
### Variant 1 (using `LOWER` function):
```sql
SELECT TABLE_SCHEMA,
       TABLE_NAME
FROM information_schema.TABLES
WHERE LOWER(TABLE_NAME) LIKE LOWER('k%')
ORDER BY TABLE_NAME
LIMIT 1000;
```

### Variant 2 (using _case insensitive_ collation dynamically):
```sql
SELECT TABLE_SCHEMA,
       TABLE_NAME
FROM information_schema.TABLES
WHERE TABLE_NAME COLLATE utf8_unicode_ci LIKE 'k%'
ORDER BY TABLE_NAME
LIMIT 1000;
```
ℹ️ NOTE  
_Using the above collation works for MySQL, for other databases the selection of the appropriate collation should be done as needed._

---

So updating the underlying SQL used for "Open Database Object" functionality in DBeaver would be really nice.
